File: src/main/java/com/wafflestudio/draft/api/RegionApiController.java
package com.wafflestudio.draft.api;

import com.wafflestudio.draft.dto.request.GetRegionsRequest;
import com.wafflestudio.draft.dto.response.RegionResponse;
import com.wafflestudio.draft.model.Region;
import com.wafflestudio.draft.service.RegionService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ModelAttribute;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import javax.validation.Valid;
import java.util.ArrayList;
import java.util.List;

@RestController
@RequestMapping("/api/v1/region")
@RequiredArgsConstructor
public class RegionApiController {

    private final RegionService regionService;

    @GetMapping("/")
    public List<RegionResponse> getRegionsV1(@Valid @ModelAttribute GetRegionsRequest request) {
        String name = request.getName();
        if (name == null) {
            name = "";
        }
        List<Region> regions = regionService.findRegionsByName(name);
        List<RegionResponse> getRegionsResponse = new ArrayList<>();
        for (Region region : regions) {
            getRegionsResponse.add(new RegionResponse(region));
        }
        return getRegionsResponse;
    }
}

File: src/main/kotlin/com/wafflestudio/draft/api/RegionApiController.kt
package com.wafflestudio.draft.api

import com.wafflestudio.draft.dto.RegionDTO
import com.wafflestudio.draft.dto.response.ListResponse
import com.wafflestudio.draft.dto.response.PageResponse
import com.wafflestudio.draft.service.RegionService
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.ModelAttribute
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import org.springframework.data.domain.PageRequest
import javax.validation.Valid

@RestController
@RequestMapping("/api/v1/region")
class RegionApiController(
        private val regionService: RegionService
) {
    @GetMapping("/")
    fun getRegionsV1(@Valid @ModelAttribute request: RegionDTO.Request): PageResponse<RegionDTO.Summary> {
        val depth3 = request.depth3.orEmpty()
        val page = request.page
        val pageable = PageRequest.of(page, 20)

        val regions = regionService.findRegionsByDepth3(depth3, pageable)
        return PageResponse(regions)
    }

    @GetMapping("/room/")
    fun getRoomsV1(): ListResponse<RegionDTO.ResponseWithRooms> {
        val regions = regionService.getRegions().map { it.toResponseWithRooms() }
        return ListResponse(regions)
    }
}
